src/cli/argsource.rs

```diff
@@ -0,0 +1,54 @@
+/// Abstracts over reading the current process environment variables as a
+/// zero-cost abstraction to support threaded in-process testing.
+use std::env;
+
+pub trait ArgSource {
+    type Args: Iterator<Item = String>;
+
+    fn args(&self) -> Self::Args;
+}
+
+/// Implements ArgSource with `std::env::args`
+pub struct OsArgs {}
+
+impl ArgSource for OsArgs {
+    type Args = env::Args;
+    fn args(&self) -> Self::Args {
+        env::args()
+    }
+}
+
+/// Helper for ArgSource over `Vec<String>`
+pub struct VecArgs {
+    v: Vec<String>,
+    i: usize,
+}
+
+impl From<&Vec<String>> for VecArgs {
+    fn from(source: &Vec<String>) -> Self {
+        let v = source.clone();
+        VecArgs { v, i: 0 }
+    }
+}
+
+impl Iterator for VecArgs {
+    type Item = String;
+    fn next(&mut self) -> Option<String> {
+        if self.i == self.v.len() {
+            return None;
```

Comment:
If you want to retain this implementation then it'd be right to also `impl FusedIterator` for this because you are fused.  Not that we rely on this but I think it does allow theoretically some optimisation.

Reply:
it allows optimising .fuse(), so sure, we can; not an issue in practice I think